src/cli.rs

```diff
@@ -1,3 +1,354 @@
+        println!("{}", message);
+        Ok(())
+    }
+    fn read(&self) -> Result<String, String> {
+        let mut buffer = String::new();
+        match std::io::stdin().read_line(&mut buffer) {
+            Ok(_) => return Ok(buffer),
+            Err(e) => return Err(e.to_string()),
+        };
+    }
+}
+
+struct CommandRunner<T: UserInterface, U: DeviceOperations> {
+    console: T,
+    operations: U,
+}
+
+impl<T: UserInterface, U: DeviceOperations> CommandRunner<T, U> {
+    fn run(&mut self, args: Vec<String>) {
+        if args.len() < 2 {
+            self.display_invalid_command();
+            return;
+        }
+
+        match args[1].as_str() {
+            "help" => self.display_help(),
+            "--version" | "-v" => self.display_version(),
+            "device" => self.run_device_command(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_message(&mut self, message: &String) {
+        let _ = self.console.write(message);
+    }
+
+    fn read_string(&mut self) -> Result<String, String> {
+        self.console.read()
+    }
+
+    fn read_number(&mut self) -> Result<i32, String> {
+        let input = self.console.read()?;
+        match input.trim().parse::<i32>() {
+            Ok(number) => Ok(number),
+            Err(_) => Err("Invalid number".to_string()),
+        }
+    }
+
+    fn display_help(&mut self) {
+        let _ = self.display_message(&HELP.to_string());
+    }
+
+    fn display_version(&mut self) {
+        let _ = self.display_message(&VERSION.to_string());
+    }
+
+    fn display_invalid_command(&mut self) {
+        let _ = self.display_message(&INVALID_COMMAND.to_string());
+    }
+
+    fn run_device_command(&mut self, args: Vec<String>) {
+        if args.len() < 3 {
+            self.display_invalid_command();
+            return;
+        }
+
+        match args[2].as_str() {
+            "ls" | "list" => self.display_device_list(),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_device_list(&mut self) {
+        let _ = self.display_message(&"Device list:".to_string());
+        let devices = self.operations.list();
+        match devices {
+            Ok(devices) => {
+                for device in devices {
+                    let _ = self.display_message(&format!("Device: {}", device.get_name()));
+                }
+            }
+            Err(e) => {
+                let _ = self.display_message(&e);
+            }
+        }
+    }
+}
+
+pub fn run(args: Vec<String>) {
+    let standard_path_provider = StandardPathProvider {};
+    let local_file_storage = LocalFileStorage::new(
+        &standard_path_provider,
+        &StandardFileSystem {},
+        DEFAULT_CONFIG,
+    );
+    let operations = Operations::new(Box::new(local_file_storage));
+    let mut command_runner = CommandRunner {
```

Comment:
You will probably need to call `register_device_factory` on concrete Operations struct. It is the part that is doing the injection of the factories with all concrete types.